File: logfalcon/fc/detector.py
"""FC detection and identification via MSP handshake."""

from __future__ import annotations

import logging
from dataclasses import dataclass

from logfalcon.msp.client import MSPClient, MSPError
from logfalcon.msp.constants import (
    BLACKBOX_DEVICE_NONE,
    BLACKBOX_DEVICE_SDCARD,
    BTFL_VARIANT,
)

log = logging.getLogger(__name__)


@dataclass
class FCInfo:
    api_major: int
    api_minor: int
    variant: bytes  # e.g. b'BTFL'
    uid: str  # hex string, e.g. "12ab34cdef..."
    blackbox_device: int


class FCDetectionError(Exception):
    pass


class FCNotBetaflight(FCDetectionError):
    pass


class FCSDCardBlackbox(FCDetectionError):
    """FC uses SD card for blackbox — must be read directly."""

    pass


class FCBlackboxEmpty(FCDetectionError):
    """Flash is already empty — nothing to sync."""

    pass


def detect_fc(client: MSPClient) -> FCInfo:
    """Run MSP handshake, verify BTFL, return FC info.

    Raises:
        FCNotBetaflight: Variant != BTFL
        FCSDCardBlackbox: Blackbox device is SD card
        FCDetectionError: Other identification failure
    """
    try:
        major, minor = client.get_api_version()
        log.info('MSP API version: %d.%d', major, minor)
    except MSPError as exc:
        raise FCDetectionError(f'MSP API_VERSION failed: {exc}') from exc

    try:
        variant = client.get_fc_variant()
        log.info('FC variant: %r', variant)
    except MSPError as exc:
        raise FCDetectionError(f'MSP FC_VARIANT failed: {exc}') from exc

    if variant[:4] != BTFL_VARIANT:
        raise FCNotBetaflight(f'Expected BTFL variant, got {variant!r}')

    uid = 'unknown'
    try:
        uid = client.get_uid()
        log.info('FC UID: %s', uid)
    except MSPError:
        log.warning("Could not read FC UID, using 'unknown'")

    blackbox_device = BLACKBOX_DEVICE_NONE
    try:
        bb_cfg = client.get_blackbox_config()
        blackbox_device = bb_cfg.get('device', BLACKBOX_DEVICE_NONE)
        log.info('Blackbox device type: %d', blackbox_device)
    except MSPError as exc:
        log.warning('Could not read BLACKBOX_CONFIG: %s', exc)

    if blackbox_device == BLACKBOX_DEVICE_SDCARD:
        raise FCSDCardBlackbox(
            'FC uses SD card for blackbox — remove the FC SD card and read it directly'
        )

    return FCInfo(
        api_major=major,
        api_minor=minor,
        variant=variant,
        uid=uid,
        blackbox_device=blackbox_device,
    )

File: internal/fc/detector.go
// Package fc provides flight-controller detection via MSP handshake.
package fc

import (
	"fmt"
	"log"

	"github.com/proeugene/logfalcon/internal/msp"
)

// FCInfo holds identification info from the MSP handshake.
type FCInfo struct {
	APIMajor       int
	APIMinor       int
	Variant        string // "BTFL" or "INAV"
	UID            string // hex string
	BlackboxDevice int
}

// DetectionError indicates a generic MSP identification failure.
type DetectionError struct {
	Message string
}

func (e *DetectionError) Error() string { return e.Message }

// NotSupportedError indicates the FC variant is not Betaflight or iNav.
type NotSupportedError struct {
	Variant string
}

func (e *NotSupportedError) Error() string {
	return fmt.Sprintf("unsupported FC variant %q", e.Variant)
}

// SDCardError indicates the FC uses an SD card for blackbox storage.
type SDCardError struct{}

func (e *SDCardError) Error() string {
	return "FC uses SD card for blackbox — remove the FC SD card and read it directly"
}

// BlackboxEmptyError indicates the flash is already empty.
type BlackboxEmptyError struct{}

func (e *BlackboxEmptyError) Error() string {
	return "flash is already empty — nothing to sync"
}

// MSPClient is the interface the detector needs from the MSP client.
// Using an interface allows testing without a real serial port.
type MSPClient interface {
	GetAPIVersion() (int, int, error)
	GetFCVariant() (string, error)
	GetUID() (string, error)
	GetBlackboxConfig() (int, error)
}

// Detect runs the MSP handshake and returns FC info.
//
// Steps (matching the Python detect_fc):
//  1. GetAPIVersion — log result; wrap errors as DetectionError
//  2. GetFCVariant  — check against SupportedVariants; NotSupportedError if unknown
//  3. GetUID        — use "unknown" on error
//  4. GetBlackboxConfig — BTFL queries MSP; INAV skips (assumes flash)
//  5. SDCard device → SDCardError
func Detect(client MSPClient) (*FCInfo, error) {
	// 1. API version
	major, minor, err := client.GetAPIVersion()
	if err != nil {
		return nil, &DetectionError{Message: fmt.Sprintf("MSP API_VERSION failed: %v", err)}
	}
	log.Printf("MSP API version: %d.%d", major, minor)

	// 2. FC variant
	variant, err := client.GetFCVariant()
	if err != nil {
		return nil, &DetectionError{Message: fmt.Sprintf("MSP FC_VARIANT failed: %v", err)}
	}
	log.Printf("FC variant: %q", variant)

	if len(variant) > 4 {
		variant = variant[:4]
	}
	if !msp.SupportedVariants[variant] {
		return nil, &NotSupportedError{Variant: variant}
	}

	// 3. UID — best effort
	uid := "unknown"
	if u, err := client.GetUID(); err == nil {
		uid = u
		log.Printf("FC UID: %s", uid)
	} else {
		log.Printf("Could not read FC UID, using 'unknown'")
	}

	// 4. Blackbox config
	bbDevice := msp.BlackboxDeviceNone
	if variant == msp.BTFLVariant {
		deviceType, err := client.GetBlackboxConfig()
		if err != nil {
			log.Printf("Could not read BLACKBOX_CONFIG: %v", err)
		} else {
			bbDevice = deviceType
			log.Printf("Blackbox device type: %d", bbDevice)
		}
	} else {
		// iNav deprecated MSP_BLACKBOX_CONFIG — assume flash until
		// DATAFLASH_SUMMARY proves otherwise.
		bbDevice = msp.BlackboxDeviceFlash
		log.Printf("Non-Betaflight FC — skipping BLACKBOX_CONFIG, assuming flash")
	}

	// 5. SD card → error
	if bbDevice == msp.BlackboxDeviceSDCard {
		return nil, &SDCardError{}
	}

	return &FCInfo{
		APIMajor:       major,
		APIMinor:       minor,
		Variant:        variant,
		UID:            uid,
		BlackboxDevice: bbDevice,
	}, nil
}
